cpp/include/raft/spatial/knn/ivf_flat.cuh

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "detail/ivf_flat_build.cuh"
+#include "detail/ivf_flat_search.cuh"
+#include "ivf_flat_types.hpp"
+
+#include <raft/core/handle.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace raft::spatial::knn::ivf_flat {
+
+/**
+ * @brief Build the index from the dataset for efficient search.
+ *
+ * NB: Currently, the following distance metrics are supported:
+ *   L2Expanded
+ *   L2Unexpanded
+ *   InnerProduct
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices in the source dataset
+ *
+ * @param handle
+ * @param params configure the index building
+ * @param[in] dataset a device pointer to a row-major matrix [n_rows, dim]
+ * @param n_rows the number of samples
+ * @param dim the dimensionality of the data
+ * @param stream
+ *
+ * @return the constructed ivf-flat index
+ */
+template <typename T, typename IdxT = uint32_t>
+inline auto build(const handle_t& handle,
+                  const index_params& params,
```

Comment:
I notice there's no way to specify the metric here or in the index_params. We should probably add that somewhere, even if we just limit it to euclidean to start, so that we don't have to break the API when add support for more of them.